### PR TITLE
[BUG-1865] Added serialization support for typescript on complex object headers.

### LINKED ISSUE #1865: [BUG] [Typescript] [Axios] Header does not support complex-objects
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] What's the version of OpenAPI Generator used?
- [ ] Have you search for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

When using a header that is a complex object like defined here https://swagger.io/specification/#headerObject the serialization is missing here:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-axios/api.mustache#L175
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-axios/api.mustache#L180
this should be done like here:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-axios/api.mustache#L226  

##### openapi-generator version

master

##### OpenAPI declaration file content or url

https://gist.github.com/sruehl/9919a4f8aab6ef9bb1ca47cf90b3c6e5

##### Related issues/PRs

#1874

##### Suggest a fix

implement serialization like here https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-axios/api.mustache#L226

### PR DESCRIPTION
**Pull request by sruehl**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

When using headers which are defined as complex objects we need to render them as json

fixes #1865 

cc: @TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @nicokoenig (2018/09) @topce (2018/10)

### CODE REVIEW

File: samples/client/petstore/typescript-axios/builds/with-complex-headers/api.ts [hunk 1]
```diff
@@ -0,0 +1,2021 @@
...
+            // oauth required
+            if (configuration && configuration.accessToken) {
+				const localVarAccessTokenValue = typeof configuration.accessToken === 'function'
+					? configuration.accessToken("petstore_auth", ["write:pets", "read:pets"])
+					: configuration.accessToken;
+                localVarHeaderParameter["Authorization"] = "Bearer " + localVarAccessTokenValue;
+            }
+
+            if (header1 !== undefined && header1 !== null) {
+                let mapped = (<any>"Pet" !== "string") ? JSON.stringify(header1 || {}) : (header1 || "")
```
Comment (sruehl) [inline samples/client/petstore/typescript-axios/builds/with-complex-headers/api.ts]: simple serialization

File: modules/openapi-generator/src/main/resources/typescript-axios/api.mustache [hunk 2]
```diff
@@ -172,12 +172,14 @@ export const {{classname}}AxiosParamCreator = function (configuration?: Configur
...
             if ({{paramName}}) {
-                localVarHeaderParameter['{{baseName}}'] = {{paramName}}.join(COLLECTION_FORMATS["{{collectionFormat}}"]);
+                let mapped = {{paramName}}.map(value => (<any>"{{{dataType}}}" !== "Array<string>") ? JSON.stringify(value || {}) : (value || ""))
+                localVarHeaderParameter['{{baseName}}'] = mapped.join(COLLECTION_FORMATS["{{collectionFormat}}"]);
             }
             {{/isListContainer}}
             {{^isListContainer}}
             if ({{paramName}} !== undefined && {{paramName}} !== null) {
-                localVarHeaderParameter['{{baseName}}'] = String({{paramName}});
+                let mapped = (<any>"{{dataType}}" !== "string") ? JSON.stringify({{paramName}} || {}) : ({{paramName}} || "")
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: @sruehl thanks for the PR. Can we use something like `{{#isString}} ... {{/isString}}` to check instead?

Comment (sruehl) [reply modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: to be honest I have no idea if its possible. What does this isString macro do?

Comment (wing328) [reply modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: `isString` mustache tag is true if the data type is "string" 

(I noticed you compared datatype to "string" in the TS code)

Comment (sruehl) [reply modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: the code is the same as the one found [here](https://github.com/OpenAPITools/openapi-generator/blob/ccde3460805c1048067e241bc6a057ce92ce6a42/modules/openapi-generator/src/main/resources/typescript-axios/api.mustache#L228), so this class might need some more work

Comment (sruehl) [reply modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: btw. is there an example where to use it like if else?

Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: https://mustache.github.io/mustache.5.html#Inverted-Sections

Comment (sruehl) [reply modules/openapi-generator/src/main/resources/typescript-axios/api.mustache]: I changed it to the proposed macro

**Comment by macjohnny** [comment]
@sruehl could you fix the conflicts?

**Comment by sruehl** [comment]
@macjohnny I merged the upstream master and fixed the merge after that. Additionally I re-ran the  `bin/typescript-axios-petstore-with-complex-headers.sh`  which produced no diff. LGTM

File: modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache [hunk 3]
```diff
@@ -133,12 +133,18 @@ export const {{classname}}AxiosParamCreator = function (configuration?: Configur
     {{#headerParams}}
             {{#isListContainer}}
             if ({{paramName}}) {
-                localVarHeaderParameter['{{baseName}}'] = {{paramName}}.join(COLLECTION_FORMATS.{{collectionFormat}});
+                let mapped = {{paramName}}.map(value => (<any>"{{{dataType}}}" !== "Array<string>") ? JSON.stringify(value || {}) : (value || ""));
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache]: if we have an `Array<number>` with `[0, 2, 3]`, wouldn't this be serialized to `{},2,3`?

File: bin/typescript-axios-petstore-all.sh [hunk 4]
```diff
@@ -3,5 +3,6 @@
 ./bin/typescript-axios-petstore-target-es6.sh
 ./bin/typescript-axios-petstore-with-npm-version.sh
 ./bin/typescript-axios-petstore-with-npm-version-and-separate-models-and-api.sh
+./bin/typescript-axios-petstore-with-complex-headers.sh
```
Comment (macjohnny) [inline bin/typescript-axios-petstore-all.sh]: please add this to the windows scripts in `bin/windows/typescript-axios-petstore-all.bat`

Comment (sruehl) [reply bin/typescript-axios-petstore-all.sh]: done

File: modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache [hunk 5]
```diff
@@ -133,12 +133,18 @@ export const {{classname}}AxiosParamCreator = function (configuration?: Configur
...
+                localVarHeaderParameter['{{baseName}}'] = mapped.join(COLLECTION_FORMATS["{{collectionFormat}}"]);
             }
             {{/isListContainer}}
             {{^isListContainer}}
             if ({{paramName}} !== undefined && {{paramName}} !== null) {
+                {{#isString}}
                 localVarHeaderParameter['{{baseName}}'] = String({{paramName}});
+                {{/isString}}
+                {{^isString}}
+                localVarHeaderParameter['{{baseName}}'] = String(JSON.stringify({{paramName}} || {}));
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache]:
```suggestion
                localVarHeaderParameter['{{baseName}}'] = String(JSON.stringify({{paramName}}));
```

File: modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache (hunk 3 above)
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache]:
```suggestion
                let mapped = {{paramName}}.map(value => (<any>"{{{dataType}}}" !== "Array<string>") ? JSON.stringify(value) : (value || ""));
```

Comment (sruehl) [reply modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache]: maybe, do you have a suggested fix?

Comment (sruehl) [reply modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache]: did your suggestion fix this too? Im not sure right now, so many months have passed ;)

Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache]: yes, this should be fixed with my suggestion

**Comment by macjohnny** [comment]
@sruehl can you re-generate the samples, please?

**Comment by sruehl** [comment]
@macjohnny done (as allways forgot to rerun` maven install`)

**Comment by sruehl** [comment]
the failing build is because of kotlin, seems like this is not related to this PR...

**Comment by macjohnny** [comment]
@sruehl can you please merge the current master? the CI build failure should be fixed

**Comment by wing328** [comment]
@sruehl thanks for the PR, which has been included in the 4.1.0 release: https://twitter.com/oas_generator/status/1160000504455319553

**Pull request merged**